Incorrect kind check error

## reproduction steps

using Scala 2.13.3, compile the following code

```scala
abstract class Expr[-S[-l[-_, +_], +f[-_, +_]], -I, +O] extends Unindexed[S] {
  def fold[F[-_, +_]](matcher: S[Expr[S, -*, +*], F]): F[I, O]

  final override def foldU[F[-_, +_]]( matcher: S[Unindexed.Inner[S, -*, +*], F]): F[Nothing, Any] = 
     fold(matcher)
}


trait Unindexed[-S[-l[-_, +_], +f[-_, +_]]] {
  def foldU[F[-i, +o]](matcher: S[Unindexed.Inner[S, -*, +*], F]): F[Nothing, Any]
}

object Unindexed {
  type Inner[-S[-l[-_, +_], +f[-_, +_]], -I, +O] = Unindexed[S]
}
```

## problem
Compilation fails with error 
> type F takes type parameters
 
at the definition of the  `Unindexed.foldU`

Change the order of two definitions to 
```scala
trait Unindexed[-S[-l[-_, +_], +f[-_, +_]]] {
  def foldU[F[-i, +o]](matcher: S[Unindexed.Inner[S, -*, +*], F]): F[Nothing, Any]
}

object Unindexed {
  type Inner[-S[-l[-_, +_], +f[-_, +_]], -I, +O] = Unindexed[S]
}

abstract class Expr[-S[-l[-_, +_], +f[-_, +_]], -I, +O] extends Unindexed[S] {
  def fold[F[-_, +_]](matcher: S[Expr[S, -*, +*], F]): F[I, O]

  final override def foldU[F[-_, +_]](matcher: S[Unindexed.Inner[S, -*, +*], F]): F[Nothing, Any] = 
     fold(matcher)
}
```
makes error disappear

Minimised:

```scala
$ m t12187.scala
object Z {
   type Baz[S[_[_], _[_]], _] = Bar[S]
  trait Foo[S[_[_], _[_]]   ] extends Bar[S] { def m[F[_]](x: S[Baz[S, *], F]): Unit }
  trait Bar[S[_[_], _[_]]   ]                { def m[F[_]](x: S[Baz[S, *], F]): Unit }
}
$ scalac -2.13.4 -Xplugin:$(cs fetch -p org.typelevel:kind-projector_2.13.4:0.11.2) t12187.scala
t12187.scala:4: error: type F takes type parameters
  trait Bar[S[_[_], _[_]]   ]                { def m[F[_]](x: S[Baz[S, *], F]): Unit }
                                                                           ^
1 error
```

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -5535,11 +5535,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       }
 
       def typedAppliedTypeTree(tree: AppliedTypeTree) = {
-        val tpt        = tree.tpt
-        val args       = tree.args
-        val tpt1       = typed1(tpt, mode | FUNmode | TAPPmode, WildcardType)
-        def isPoly     = tpt1.tpe.isInstanceOf[PolyType]
-        def isComplete = tpt1.symbol.rawInfo.isComplete
+        val tpt = tree.tpt
+        val args = tree.args
+        val tpt1 = typed1(tpt, mode | FUNmode | TAPPmode, WildcardType)
 
         if (tpt1.isErrorTyped) {
           tpt1
@@ -5547,6 +5545,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           AppliedTypeNoParametersError(tree, tpt1.tpe)
         } else {
           val tparams = tpt1.symbol.typeParams
+          val isComplete = tpt1.symbol.rawInfo.isComplete
 
           if (sameLength(tparams, args)) {
             // @M: kind-arity checking is done here and in adapt, full kind-checking is in checkKindBounds (in Infer)
@@ -5594,6 +5593,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             }
             val original = treeCopy.AppliedTypeTree(tree, tpt1, args1)
             val result = TypeTree(appliedType(tpt1.tpe, argtypes)) setOriginal original
+            val isPoly = tpt1.tpe.isInstanceOf[PolyType]
             if (isPoly) // did the type application (performed by appliedType) involve an unchecked beta-reduction?
               TypeTreeWithDeferredRefCheck() { () =>
                 // wrap the tree and include the bounds check -- refchecks will perform this check (that the beta reduction was indeed allowed) and unwrap

--- a/test/files/pos/t12187.scala
+++ b/test/files/pos/t12187.scala
@@ -1,0 +1,8 @@
+object Test {
+  trait Foo[S[_[_], _[_]]] extends Bar[S] {
+    def m[F[_]](x: S[({ type G[A] = Bar[S] })#G, F]): Unit
+  }
+  trait Bar[S[_[_], _[_]]] {
+    def m[F[_]](x: S[({ type G[A] = Bar[S] })#G, F]): Unit
+  }
+}

